src/lib/storage/vector_compression/bitpacking/bitpacking_compressor.cpp

```diff
@@ -0,0 +1,46 @@
+#include "bitpacking_compressor.hpp"
+
+#include <algorithm>
+#include <cmath>
+
+#include "bitpacking_vector_type.hpp"
+
+namespace opossum {
+
+class BitPackingVector;
+
+std::unique_ptr<const BaseCompressedVector> BitPackingCompressor::compress(const pmr_vector<uint32_t>& vector,
+                                                                           const PolymorphicAllocator<size_t>& alloc,
+                                                                           const UncompressedVectorInfo& meta_info) {
+  const auto max_element_it = std::max_element(vector.cbegin(), vector.cend());
+
+  auto required_bits = 1u;
+  if (max_element_it != vector.cend() && *max_element_it != 0) {
+    // add 1 to max_value because log2(1) = 0 but we need one bit to represent it.
+    required_bits = static_cast<uint32_t>(std::ceil(log2(*max_element_it + 1u)));
+  }
+
+  auto data = pmr_compact_vector(required_bits, vector.size(), alloc);
+
+  /**
+   * The compact::vector does not zero initialize its memory, which leads to non-reproducible tests that
+   * use the binary writer. The compact::vector allocates memory aligned to its word size and due to this,
+   * some memory at the end is not always overwritten. Hence, fill the internal memory with zeroes.
+   * Unfortunately, compact::vector does not give us a better interface. Therefore, we proceed to work with
+   * a raw pointer to the compact::vector's internal memory and fill it with zeroes. For this, data.bytes() 
+   * gives the number of allocated bytes for the internal memory 
+   * (word-aligned, see bitpacking_vector_type.hpp). 
+   *
+   */
+  std::fill_n(data.get(), data.bytes() / sizeof(*data.get()), 0ULL);
```

Comment:
You assume that we still have the word size of `uint64_t`, right? If somebody changes that to `uint32_t`, this could cause problems. Can't we do `data.bytes() / sizeof(uint64_t), uint64_t{0}`?

Reply:
I'm not sure if I follow you. I think the following happens:

- for the word size uint64_t: 
   - data.get() will be a uint64_t*
   - sizeof(*data.get()) = 8
   - data.bytes() / sizeof(*data.get()) will be the number of uint64_t values that fit into the given number of bytes
   - 0ULL will be an unsigned long long (at least 64 bits)
   
   
- if we change the word size to uint32_t for example: 
   - data.get() will be a uint32_t*
   - sizeof(*data.get()) = 4
   - data.bytes() / sizeof(*data.get()) will be the number of uint32_t values that fit into the given number of bytes
   - 0ULL will be an unsigned long long (at least 64 bits) but it does not matter since it will be implicitly converted to a uint32_t

Reply:
My point is that if somebody changes the `using compact_vector` definition, this code will no longer work since you'll write too many large integers. Maybe DebugAssert that `sizeof(data.get)` is uint64_t and then also write a `uint64_t{0}` (that's just for consistency and readability)?

Reply:
Okay, I did that